internal/pkg/composable/providers/kubernetessecrets/kubernetes_secrets.go

```diff
@@ -46,38 +55,152 @@ func ContextProviderBuilder(logger *logger.Logger, c *config.Config, managed boo
 		return nil, errors.New(err, "failed to unpack configuration")
 	}
 	return &contextProviderK8sSecrets{
-		logger: logger,
-		config: &cfg,
+		logger:       logger,
+		config:       &cfg,
+		secretsCache: make(map[string]*secretsData),
 	}, nil
 }
 
 func (p *contextProviderK8sSecrets) Fetch(key string) (string, bool) {
-	// key = "kubernetes_secrets.somenamespace.somesecret.value"
+	return p.getFromCache(key)
+}
+
+// Run initializes the k8s secrets context provider.
+func (p *contextProviderK8sSecrets) Run(ctx context.Context, comm corecomp.ContextProviderComm) error {
+	client, err := getK8sClientFunc(p.config.KubeConfig, p.config.KubeClientOptions)
+	if err != nil {
+		p.logger.Debugf("Kubernetes_secrets provider skipped, unable to connect: %s", err)
+		return nil
+	}
 	p.clientMx.Lock()
-	client := p.client
+	p.client = client
 	p.clientMx.Unlock()
-	if client == nil {
-		return "", false
+	go p.updateSecrets(ctx)
+
+	<-comm.Done()
+
+	p.clientMx.Lock()
+	p.client = nil
+	p.clientMx.Unlock()
+	return comm.Err()
+}
+
+func getK8sClient(kubeconfig string, opt kubernetes.KubeClientOptions) (k8sclient.Interface, error) {
+	return kubernetes.GetKubernetesClient(kubeconfig, opt)
+}
+
+// Update the secrets in the cache every TTL minutes
+func (p *contextProviderK8sSecrets) updateSecrets(ctx context.Context) {
+	timer := time.NewTimer(p.config.TTLUpdate)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-timer.C:
+			p.updateCache()
+			timer.Reset(p.config.TTLUpdate)
+		}
 	}
+}
```

Comment:
> General question: why do we update all secrets periodically instead of adding/refreshing entries based on access?

We cannot do that, because otherwise we would need to make a request to the API Server every time we want to access the value. That is the [issue](https://github.com/elastic/elastic-agent/issues/3594) that led to this PR.

> If we set a TTL for deletion (basically the time we consider a cache value "not stale") and in case of cache miss we fetch the specific secret value woudn't that be more efficient ?

You mean to just update the secret after TTL delete has passed? TTL delete is quite a long duration.

> it's just to understand the rationale behind having a cache where (almost) all entries get refreshed at once: do we have an idea of what the access pattern of secrets would be ?

I am not sure, this a feature request. In the issue I [linked](https://github.com/elastic/elastic-agent/issues/3594) we discuss the possible workarounds. In the future, this cache will likely be replaced by a watcher, but we need to solve [this](https://github.com/elastic/beats/issues/34717) first.